README.rst
QBroker is short for Quick Access Broker.

QBroker is Python library which provides an easy-to-use interface for asynchronous, debuggable RPC via AMQP.
It uses Python 3.4 asyncio features and provides wrappers for use with native threads.

Example scripts, including an event/RPC monitor, are provided.

The QBroker data format is straightforward. Writing an adapter in a different programming language
is straightforward, given JSON and AMQP libraries. You can add codecs for your classes, or replace
JSON with another marshalling system.

QBroker supports the time-to-live and dead-lettering extensions of RabbitMQ, but does not depend on them.
Its support for monitoring does not rely on RabbitM's "firehose" tracing feature.

QBroker supports (as-yet rudimentary) introspection.


debian/changelog
qbroker (0.31.1-3) unstable; urgency=medium

  * Fix test

 -- Matthias Urlichs <matthias@urlichs.de>  Tue, 27 Nov 2018 09:58:01 +0100

qbroker (0.31.1-2) unstable; urgency=medium

  * avoid 'async' keyword

 -- Matthias Urlichs <matthias@urlichs.de>  Tue, 27 Nov 2018 09:50:32 +0100

qbroker (0.31.1-1) unstable; urgency=medium

  * Drop support for gevent.

 -- Matthias Urlichs <matthias@urlichs.de>  Tue, 27 Nov 2018 09:49:37 +0100

qbroker (0.30.8-2) unstable; urgency=medium

  * Ignore pytest cache dir

 -- Matthias Urlichs <matthias@urlichs.de>  Tue, 27 Nov 2018 09:41:50 +0100

qbroker (0.30.8-1) unstable; urgency=medium

  * Don't mark the json+repr codec as default

 -- Matthias Urlichs <matthias@urlichs.de>  Thu, 14 Sep 2017 10:38:17 +0200

qbroker (0.30.7-1) unstable; urgency=medium

  * Add json/obj+repr codec
  * Extend debugger

 -- Matthias Urlichs <matthias@urlichs.de>  Wed, 13 Sep 2017 23:44:47 +0200

qbroker (0.30.6-1) unstable; urgency=medium

  * Forgot to think …

 -- Matthias Urlichs <matthias@urlichs.de>  Wed, 13 Sep 2017 14:48:03 +0200

qbroker (0.30.5-1) unstable; urgency=medium

  * Prevent restarts during closing.

 -- Matthias Urlichs <matthias@urlichs.de>  Fri, 18 Aug 2017 10:11:18 +0200

qbroker (0.30.4-1) unstable; urgency=medium

  * Error handling fix

 -- Matthias Urlichs <matthias@urlichs.de>  Wed, 16 Aug 2017 16:50:14 +0200

qbroker (0.30.3-1) unstable; urgency=medium

  * Wrong variable

 -- Matthias Urlichs <matthias@urlichs.de>  Tue, 15 Aug 2017 18:49:11 +0200

qbroker (0.30.2-5) smurf; urgency=medium

  * Travis: Python 3.4 does no longer work
    because pytest_asyncio now requires py3.5.

 -- Matthias Urlichs <matthias@urlichs.de>  Thu, 10 Aug 2017 12:06:53 +0200

qbroker (0.30.2-4) smurf; urgency=medium

  * Travis: fix pyyaml requirement

 -- Matthias Urlichs <matthias@urlichs.de>  Thu, 10 Aug 2017 11:55:27 +0200

qbroker (0.30.2-3) smurf; urgency=medium

  * Increment the version number, dammit

 -- Matthias Urlichs <matthias@urlichs.de>  Thu, 10 Aug 2017 11:42:12 +0200

qbroker (0.30.2-2) smurf; urgency=medium

  * Allow specifying a codec for replies

    This involves using CC_MSG and setting msg.codec to the codec you want

 -- Matthias Urlichs <matthias@urlichs.de>  Thu, 10 Aug 2017 11:30:02 +0200

qbroker (0.30.1-1) smurf; urgency=medium

  * Allow specifying a codec when sending a message.

    TODO: doing this when replying

  * Deprecate passing random keywords into msg() and alert().
    Use a dict(…) argument, please.
    This also removes the underscores from these methods' parameters.

 -- Matthias Urlichs <matthias@urlichs.de>  Thu, 10 Aug 2017 11:17:05 +0200

qbroker (0.21.0-1) smurf; urgency=medium

  * Add a basic debugger back-end

 -- Matthias Urlichs <matthias@urlichs.de>  Wed, 19 Jul 2017 14:33:29 +0200

qbroker (0.20.0-1) smurf; urgency=medium

  * format_dt: accept timestamps, drop error 

 -- Matthias Urlichs <matthias@urlichs.de>  Sun, 16 Jul 2017 18:14:35 +0200

qbroker (0.19.5-1) smurf; urgency=medium

  * Fix compatibility with non-qbroker (non-RPC) messages

 -- Matthias Urlichs <matthias@urlichs.de>  Tue, 04 Jul 2017 06:11:38 +0200

qbroker (0.19.4-1) smurf; urgency=medium

  * Work around strange coded error messages

 -- Matthias Urlichs <matthias@urlichs.de>  Sun, 30 Apr 2017 16:09:41 +0200

qbroker (0.19.3-2) smurf; urgency=medium

  * update .gitignore
  * kill test.log file

 -- Matthias Urlichs <matthias@urlichs.de>  Sun, 30 Apr 2017 15:29:36 +0200

qbroker (0.19.3-1) smurf; urgency=medium

  * Merge restart

 -- Matthias Urlichs <matthias@urlichs.de>  Sun, 30 Apr 2017 15:12:43 +0200

qbroker (0.19.2-2) unstable; urgency=medium

  * Fix testing

 -- Matthias Urlichs <matthias@urlichs.de>  Sun, 30 Apr 2017 15:02:55 +0200

qbroker (0.19.2-1) unstable; urgency=medium

  * Fix naming durable alert queues

 -- Matthias Urlichs <matthias@urlichs.de>  Sun, 30 Apr 2017 14:59:37 +0200

qbroker (0.19.1-1) unstable; urgency=medium

  * fix tests

 -- Matthias Urlichs <matthias@urlichs.de>  Mon, 05 Dec 2016 16:54:50 +0100

qbroker (0.19.0-1) unstable; urgency=medium

  * don't double-register the UUID channel
  * Add hidden server mode
  * fix server listing example
  * add firehose example

 -- Matthias Urlichs <matthias@urlichs.de>  Mon, 05 Dec 2016 14:21:35 +0100

qbroker (0.18.4-1) unstable; urgency=medium

  * Use the standard event loop in Run.

 -- Matthias Urlichs <matthias@urlichs.de>  Mon, 05 Dec 2016 11:31:10 +0100

qbroker (0.18.3-1) unstable; urgency=medium

  * Transmit Python errors natively. For real.

 -- Matthias Urlichs <smurf@noris.de>  Wed, 23 Nov 2016 12:26:32 +0100

qbroker (0.18.2-3) unstable; urgency=medium

  * require six

 -- Matthias Urlichs <matthias@urlichs.de>  Wed, 09 Nov 2016 22:29:30 +0100

qbroker (0.18.2-2) unstable; urgency=medium

  * missing import

 -- Matthias Urlichs <matthias@urlichs.de>  Wed, 09 Nov 2016 21:50:16 +0100

qbroker (0.18.1-1) unstable; urgency=medium

  * main runner: always call stop handlers

 -- Matthias Urlichs <smurf@noris.de>  Tue, 18 Oct 2016 14:52:50 +0200

qbroker (0.18.0-2) unstable; urgency=medium

  * Merge

 -- Matthias Urlichs <matthias@urlichs.de>  Mon, 03 Oct 2016 06:23:50 +0200

qbroker (0.18.0-1) unstable; urgency=medium

  * Transmit Python errors natively.

 -- Matthias Urlichs <matthias@urlichs.de>  Mon, 03 Oct 2016 05:57:54 +0200

qbroker (0.17.0-2) unstable; urgency=medium

  * missed passing the loop

 -- Matthias Urlichs <matthias@urlichs.de>  Mon, 03 Oct 2016 06:22:27 +0200

qbroker (0.17.0-1) unstable; urgency=medium

  * Add async_sync and _gevent decorators

    These wrap a synchronous / gevent function in a thread / tasklet
    so that they return an asyncio future.

 -- Matthias Urlichs <matthias@urlichs.de>  Thu, 08 Sep 2016 11:39:18 +0200

qbroker (0.16.4-3) unstable; urgency=medium

  * TODO done

 -- Matthias Urlichs <matthias@urlichs.de>  Thu, 08 Sep 2016 02:05:31 +0200

qbroker (0.16.4-2) unstable; urgency=medium

  * add TODO


qbroker/config.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# This file is part of QBroker, an easy to use RPC and broadcast
# client+server using AMQP.
#
# QBroker is Copyright © 2016-2018 by Matthias Urlichs <matthias@urlichs.de>,
# it is licensed under the GPLv3. See the file `README.rst` for details,
# including optimistic statements by the author.
#
# This paragraph is auto-generated and may self-destruct at any time,
# courtesy of "make update". The original is in ‘utils/_boilerplate.py’.
# Thus, please do not remove the next line, or insert any blank lines.
#BP

DEFAULT_CONFIG = dict(
    server=dict(
        host='localhost',
        login='guest',
        password='guest',
        virtualhost='qbroker',
        ssl=False,
    ),
    exchanges=dict(  # all are persistent
        alert='alert',  # topic: broadcast messages, may elicit multiple replies
        rpc='rpc',  # topic: RPC requests, will trigger exactly one reply
        reply='reply',  # direct: all replies go here
        dead='dead',  # fanout: dead messages
                      # (TTL expires, reject, RPC/alert unrouteable, …)
    ),
    queues=dict(
        alert='alert_',  # plus the unit UUID. Nonpersistent.
        msg='msg.',  # plus the routing key. Persistent alerts.
        rpc='rpc_',  # plus the command name. Persistent.
        reply='reply_',  # plus the unit UUID.
        dead='dead',  # no add-on. Persistent. No TTL here!
    ),
    handlers=dict(
        dead=False,  # add a handler for dead messages
        debug=False,  # add hooks to introspect/debug the connection
    ),
    codec='DEFAULT',
    ttl=dict(
        rpc=10,  # seconds before being dead lettered, when not read
        alert=10,  # durable queues only
    ),
    limits=dict(
        rpc=dict(
            workers=10,  # parallel handlers
            queue=99,  # max nr of waiting tasks
        ),
        alert=dict(
            workers=5,  # parallel handlers
            queue=99,  # max nr of waiting tasks
        ),
    ),
    timeout=dict(
        connect=30,  # open link
        rpc=15,  # waiting for RPC reply
        poll=30,  # waiting for alert replies
        reconnect=30,  # before retrying
    ),
    retries=dict(
        connect=0,  # initially
        rpc=0,  # when timed out
        poll=0,  # if no reply
        reconnect=9999,  # retries before giving up
    ),
)
